include/derecho/core/external_group.hpp

```diff
@@ -0,0 +1,108 @@
+
+    template <rpc::FunctionTag tag, typename... Args>
+    auto p2p_send(node_id_t dest_node, Args&&... args);
+};
+
+template <typename... ReplicatedTypes>
+class ExternalGroup {
+private:
+    template <typename T, typename... Ts>
+    friend class ExternalClientCaller;
+    const node_id_t my_id;
+    std::unique_ptr<View> prev_view;
+    std::unique_ptr<View> curr_view;
+    std::unique_ptr<sst::P2PConnectionManager> p2p_connections;
+    std::mutex p2p_connections_mutex;
+    std::unique_ptr<std::map<rpc::Opcode, rpc::receive_fun_t>> receivers;
+    std::map<subgroup_id_t, std::list<rpc::PendingBase_ref>> fulfilled_pending_results;
+    std::map<subgroup_id_t, uint64_t> max_payload_sizes;
+
+    template <typename T>
+    using external_caller_index_map = std::map<uint32_t, ExternalClientCaller<T, ReplicatedTypes...>>;
+    mutils::KindMap<external_caller_index_map, ReplicatedTypes...> external_callers;
+
+    bool get_view(const node_id_t nid);
+    // tcp::socket& get_socket(node_id_t nid);
+    void clean_up();
+    volatile char* get_sendbuffer_ptr(uint32_t dest_id, sst::REQUEST_TYPE type);
+    void finish_p2p_send(node_id_t dest_id, subgroup_id_t dest_subgroup_id, rpc::PendingBase& pending_results_handle);
+
+    uint32_t get_index_of_type(const std::type_info& ti);
+
+    /** ======================== copy/paste from rpc_manager ======================== **/
+    std::atomic<bool> thread_shutdown{false};
+    std::thread rpc_thread;
+    /** p2p send and queries are queued in fifo worker */
+    std::thread fifo_worker_thread;
+    struct fifo_req {
+        node_id_t sender_id;
+        char* msg_buf;
+        uint32_t buffer_size;
+        fifo_req() : sender_id(0),
+                     msg_buf(nullptr),
+                     buffer_size(0) {}
+        fifo_req(node_id_t _sender_id,
+                 char* _msg_buf,
+                 uint32_t _buffer_size) : sender_id(_sender_id),
+                                          msg_buf(_msg_buf),
+                                          buffer_size(_buffer_size) {}
+    };
+    std::queue<fifo_req> fifo_queue;
+    std::mutex fifo_queue_mutex;
+    std::condition_variable fifo_queue_cv;
+    mutils::RemoteDeserialization_v rdv;
+    void p2p_receive_loop();
+    void fifo_worker();
+    void p2p_message_handler(node_id_t sender_id, char* msg_buf, uint32_t buffer_size);
+    std::exception_ptr receive_message(const rpc::Opcode& indx, const node_id_t& received_from,
+                                       char const* const buf, std::size_t payload_size,
+                                       const std::function<char*(int)>& out_alloc);
+    /** ======================== copy/paste from rpc_manager ======================== **/
+
+public:
+    ExternalGroup(IDeserializationContext* deserialization_context = nullptr);
+    ~ExternalGroup();
+
+    template <typename SubgroupType>
+    ExternalClientCaller<SubgroupType, ReplicatedTypes...>& get_ref(uint32_t subgroup_index = 0);
```

Comment:
I think the interface of get_ref could be made a little cleaner. First, as a minor point, get_ref isn't the best name; it could at least be named get_subgroup_caller to make it clear what sort of thing you're getting a reference to. Second, putting the type names of every subgroup in the Group into the template parameters of ExternalClientCaller makes it verbose and harder to read. In a Group with three subgroup types, the user might have an ExternalClientCaller like this:

    ExternalClientCaller<FunctionServer, FunctionServer, CategorizerServer, StorageServer>& external_client;

Not only is this a very long type name, it's hard to tell from looking at it which subgroup type this external client can communicate with (eventually, you'll figure out it's the subgroup type that's printed twice). 

One easy way to improve the type name would be to make the entire ExternalGroup type a template parameter of ExternalClientCaller, instead of the ExternalGroup's parameter pack. You would declare the ExternalClientCaller like this:

    template <typename T, typename ExternalGroupType>
    class ExternalClientCaller {
    private:
        node_id_t node_id;
        subgroup_id_t subgroup_id;
        ExternalGroupType& group;
    ...
    }

and then the return type of get_ref would be `ExternalClientCaller<SubgroupType, ExternalGroup<ReplicatedTypes...>>`.  Technically, this would make the ExternalClientCaller type even more verbose, although easier to read:

    ExternalClientCaller<FunctionServer, ExternalGroup<FunctionServer, CategorizerServer, StorageServer>>& external_client;

but in practice, you could always use `decltype` to write the typename of the ExternalGroup from which you're getting the ExternalClientCaller:

    ExternalClientCaller<FunctionServer, decltype(group)>& external_client;

This is just one suggestion, though, and I welcome other ideas for how to get rid of all these extra template parameters on ExternalClientCaller.

Reply:
I hated those extra template parameters as well, but I couldn't come up with a clean workaround at that time. I think what you suggested is brilliant. I will do exactly what you suggested.